instana/propagators/base_propagator.py

```diff
@@ -41,85 +40,177 @@ class BasePropagator():
-                else:
-                    lc_key = key.lower()
-
-                if self.LC_HEADER_KEY_T == lc_key:
-                    trace_id = header_to_id(dc[key])
-                elif self.LC_HEADER_KEY_S == lc_key:
-                    span_id = header_to_id(dc[key])
-                elif self.LC_HEADER_KEY_L == lc_key:
-                    level = dc[key]
-                elif self.LC_HEADER_KEY_SYNTHETIC == lc_key:
-                    synthetic = dc[key] in ['1', b'1']
-
-                elif self.ALT_LC_HEADER_KEY_T == lc_key:
-                    trace_id = header_to_id(dc[key])
-                elif self.ALT_LC_HEADER_KEY_S == lc_key:
-                    span_id = header_to_id(dc[key])
-                elif self.ALT_LC_HEADER_KEY_L == lc_key:
-                    level = dc[key]
-                elif self.ALT_LC_HEADER_KEY_SYNTHETIC == lc_key:
-                    synthetic = dc[key] in ['1', b'1']
-
-            ctx = None
-            if trace_id is not None and span_id is not None:
-                ctx = SpanContext(span_id=span_id,
-                                  trace_id=trace_id,
-                                  level=level,
-                                  baggage={},
-                                  sampled=True,
-                                  synthetic=synthetic)
-            elif synthetic:
-                ctx = SpanContext(synthetic=synthetic)
+            ctx = self.__determine_span_context(trace_id, span_id, level, synthetic, traceparent, tracestate)
 
             return ctx
+        except Exception:
+            logger.debug("extract error:", exc_info=True)
+
+    def __determine_span_context(self, trace_id, span_id, level, synthetic, traceparent, tracestate):
+        """
+        This method determines the span context depending on a set of conditions being met
+        Detailed description of the conditions can be found here:
+        https://github.com/instana/technical-documentation/tree/master/tracing/specification#http-processing-for-instana-tracers
+        :param trace_id: instana trace id
+        :param span_id: instana span id
+        :param level: instana level
+        :param synthetic: instana synthetic
+        :return: ctx or None
+        """
+        disable_traceparent = os.environ.get("INSTANA_W3C_DISABLE_TRACE_CORRELATION", "")
+
+        ctx = None
+        if level and "correlationType" in level:
+            trace_id, span_id = [None] * 2
+
+        if traceparent and all(v is None for v in [trace_id, span_id]):
+            tp_trace_id, tp_parent_id, tp_sampled = self.__tp.get_traceparent_fields(traceparent)
+            if disable_traceparent == "":
+                ctx = SpanContext(span_id=tp_parent_id, trace_id=tp_trace_id[-16:],
+                                  level=int(level.split(",")[0]) if level else 1, baggage={}, sampled=True,
+                                  synthetic=False if synthetic is None else True)
```

Comment:
Could the actual expression value be used here?
```suggestion
                                  synthetic=synthetic is not None)
```